### DISCUSSION TITLE
Dose K3s work with Docker ( not based on docker ) now ?

### DISCUSSION THREAD

**Discussion by Alceatraz** [k3s-io/k3s#10917] [Q&A] [answered]
# For someone in future:

> WARN: Use on you own risk.

> For tidily docker ps list and registries.yaml config I choose take this risk.

Docker's containerd using systemd-v2 driver

```shell
docker info

# Cgroup Driver: systemd
# Cgroup Version: 2
```

K3s's containerd using systemd-v2 driver

```shell
k3s crictl info

# config.containerd.runtimes.runc.options.SystemdCgroup: true
```

Docker's and K3s's containerd using different main containerd

```shell
ps aux | grep containerd

# /usr/bin/containerd-shim-runc-v2 -namespace moby -address /run/containerd/containerd.sock
# /var/lib/rancher/k3s/data/???/bin/containerd-shim-runc-v2 -namespace k8s.io -address /run/k3s/containerd/containerd.sock
```

```shell
lscgroup

# ..... /system.slice/docker-0000000000.scope
# ..... /system.slice/docker-0000000000.scope
# ..... /system.slice/docker-0000000000.scope
# ..... /system.slice/kubepods-burststable.slice/kubepods-0000000000000
# ..... /system.slice/kubepods-burststable.slice/kubepods-0000000000000
# ..... /system.slice/kubepods-burststable.slice/kubepods-0000000000000
```

|        | namespace | address                             |
|--------|-----------|-------------------------------------|
| Docker | moby      | /run/containerd/containerd.sock     |
| K3s    | k8s.io    | /run/k3s/containerd/containerd.sock |

I was expecting:

```mermaid
flowchart LR
    k3s["K3s"]
    docker["Docker"]
    containerd["Containerd"]
    k3s -- " ns=k8s.io " --- containerd
    docker -- " ns=moby " --- containerd
```

But actually:

```mermaid
flowchart LR
    docker["Docker"]
    docker-containerd-shim-runc-v2["Docker-Containerd-shim-runc-v2"]
    k3s["K3s"]
    k3s-containerd-shim-runc-v2["K3s-Containerd-shim-runc-v2"]
    driver["System-CGroup-Driver"]
    k3s -- " ns=k8s.io " --- k3s-containerd-shim-runc-v2
    docker -- " ns=moby " --- docker-containerd-shim-runc-v2
    k3s-containerd-shim-runc-v2 -- " cgroup-driver " --- driver
    docker-containerd-shim-runc-v2 -- " cgroup-driver " --- driver
```

---

# Original post

From docker 1.11, Docker use containerd ( OCI/Runc ) instead DockerDaemon

About 1 years ago with debian11, I try to:

```shell
apt install docker.io
curl -sfL https://get.k3s.io | sh - # No --docker here
```

It was looking fine, But after reboot, docker service will failed to start.

Now with debian 12

- docker 20.10.24+dfsg1
- v1.30.4+k3s1 (98262b5d)

Everything works fine, Do we have officaly support Docker with K3s or just a coincidence ?

BTW: In some situations. A single node cluster lets say. I want run nginx mysql NXRM Harbor jenkins ... etc with `docker` and CI/CD with K3s. Run image registry inside cluster is a very bad idea. Use docker backend AFAIK registries.yaml not working and will let `docker ps` looks really mess.

**Comment by brandond** [comment, marked as answer]
Don't use multiple container runtimes on the same node.

If you're going to use k3s alongside docker on the same host, you should use docker as the container runtime: https://docs.k3s.io/advanced#using-docker-as-the-container-runtime

If you're trying to use k3s with the embedded containerd, that is not likely to work well alongside the host's containerd that docker uses as a backend.

**Reply by Alceatraz** [reply]
> If you're trying to use k3s with the embedded containerd, that is not likely to work well alongside the host's containerd that docker uses as a backend.

I figure it out. Thanks for reply and the hint. I was thought containerd is singletone service.
